File: github-ai-assistant.js
// github-ai-assistant.js
const fs = require('fs');
const path = require('path');

class GitHubAIAssistant {
    constructor() {
        this.supportedLanguages = new Set([
            'python', 'java', 'javascript', 'typescript', 'html',
            'css', 'cpp', 'c', 'csharp', 'go', 'rust', 'ruby'
        ]);
    }

    async analyzeRepository(repoPath) {
        const analysis = {
            totalFiles: 0,
            byLanguage: {},
            complexityScore: 0,
            issuesFound: []
        };

        try {
            const files = await this.walkDirectory(repoPath);
            
            for (const file of files) {
                const language = this.detectLanguage(file);
                
                if (language && this.supportedLanguages.has(language)) {
                    analysis.totalFiles++;
                    analysis.byLanguage[language] = (analysis.byLanguage[language] || 0) + 1;
                    
                    const complexity = await this.analyzeFileComplexity(file, language);
                    analysis.complexityScore += complexity;
                }
            }

            return analysis;
        } catch (error) {
            console.error('Ошибка анализа:', error);
            return analysis;
        }
    }

    detectLanguage(filePath) {
        const ext = path.extname(filePath).toLowerCase();
        const extensionMap = {
            '.py': 'python',
            '.java': 'java',
            '.js': 'javascript',
            '.ts': 'typescript',
            '.html': 'html',
            '.css': 'css',
            '.cpp': 'cpp',
            '.c': 'c',
            '.cs': 'csharp',
            '.go': 'go',
            '.rs': 'rust',
            '.rb': 'ruby'
        };
        return extensionMap[ext];
    }

    async analyzeFileComplexity(filePath, language) {
        try {
            const content = await fs.promises.readFile(filePath, 'utf8');
            const lines = content.split('\n');
            
            // Простая метрика сложности
            let complexity = Math.floor(lines.length / 10);
            
            // Проверка качества кода
            if (language === 'javascript') {
                if (content.includes('eval(') && content.includes('innerHTML')) {
                    this.addIssue('Возможная XSS уязвимость');
                }
            }
            
            return complexity;
        } catch (error) {
            console.error(`Ошибка анализа файла ${filePath}:`, error);
            return 0;
        }
    }

    addIssue(message) {
        this.issuesFound.push({
            message,
            severity: 'warning'
        });
    }

    walkDirectory(dir) {
        return new Promise((resolve, reject) => {
            const files = [];
            
            function walk(currentPath) {
                const items = fs.readdirSync(currentPath);
                
                for (const item of items) {
                    const fullPath = path.join(currentPath, item);
                    const stat = fs.statSync(fullPath);
                    
                    if (stat.isDirectory()) {
                        walk(fullPath);
                    } else {
                        files.push(fullPath);
                    }
                }
            }
            
            try {
                walk(dir);
                resolve(files);
            } catch (error) {
                reject(error);
            }
        });
    }
}

module.exports = GitHubAIAssistant;

File: GitHubAIAssistant.java
// GitHubAIAssistant.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.concurrent.*;

public class GitHubAIAssistant {
    private Set<String> supportedLanguages;
    private List<Map<String, String>> issuesFound;
    
    public GitHubAIAssistant() {
        this.supportedLanguages = Set.of(
            "python", "java", "javascript", "typescript", "html",
            "css", "cpp", "c", "csharp", "go", "rust", "ruby"
        );
        this.issuesFound = new ArrayList<>();
    }
    
    public Map<String, Object> analyzeRepository(String repoPath) throws IOException {
        Map<String, Object> analysis = new HashMap<>();
        analysis.put("totalFiles", 0);
        analysis.put("complexityScore", 0);
        
        Map<String, Integer> byLanguage = new HashMap<>();
        analysis.put("byLanguage", byLanguage);
        
        Files.walk(Paths.get(repoPath))
            .filter(Files::isRegularFile)
            .forEach(filePath -> {
                String language = detectLanguage(filePath.toString());
                
                if (language != null && supportedLanguages.contains(language)) {
                    byLanguage.put(language, byLanguage.getOrDefault(language, 0) + 1);
                    analysis.put("totalFiles", (Integer)analysis.get("totalFiles") + 1);
                    
                    try {
                        int complexity = analyzeFileComplexity(filePath, language);
                        analysis.put("complexityScore", (Integer)analysis.get("complexityScore") + complexity);
                    } catch (IOException e) {
                        System.err.println("Ошибка анализа файла: " + filePath);
                    }
                }
            });
        
        analysis.put("issuesFound", new ArrayList<>(issuesFound));
        return analysis;
    }
    
    private String detectLanguage(String filePath) {
        Map<String, String> extensionMap = Map.ofEntries(
            Map.entry(".py", "python"),
            Map.entry(".java", "java"),
            Map.entry(".js", "javascript"),
            Map.entry(".ts", "typescript"),
            Map.entry(".html", "html"),
            Map.entry(".css", "css"),
            Map.entry(".cpp", "cpp"),
            Map.entry(".c", "c"),
            Map.entry(".cs", "csharp"),
            Map.entry(".go", "go"),
            Map.entry(".rs", "rust"),
            Map.entry(".rb", "ruby")
        );
        
        int dotIndex = filePath.lastIndexOf('.');
        if (dotIndex > 0) {
            String extension = filePath.substring(dotIndex);
            return extensionMap.get(extension);
        }
        return null;
    }
    
    private int analyzeFileComplexity(Path filePath, String language) throws IOException {
        List<String> lines = Files.readAllLines(filePath);
        int complexity = lines.size() / 10;
        
        // Проверка качества кода
        String content = String.join("\n", lines);
        if ("java".equals(language)) {
            if (content.contains("System.out.println") && content.contains("e.printStackTrace()")) {
                addIssue("Возможно избыточное логирование", "warning");
            }
        }
        
        return complexity;
    }
    
    private void addIssue(String message, String severity) {
        Map<String, String> issue = new HashMap<>();
        issue.put("message", message);
        issue.put("severity", severity);
        issuesFound.add(issue);
    }
    
    public static void main(String[] args) {
        GitHubAIAssistant assistant = new GitHubAIAssistant();
        try {
            Map<String, Object> analysis = assistant.analyzeRepository(".");
            System.out.println("Анализ завершен:");
            System.out.println("Всего файлов: " + analysis.get("totalFiles"));
            System.out.println("По языкам: " + analysis.get("byLanguage"));
            System.out.println("Сложность: " + analysis.get("complexityScore"));
        } catch (IOException e) {
            System.err.println("Ошибка анализа репозитория: " + e.getMessage());
        }
    }
}
